File: aoc2016/day02/src/test/java/de/havox_design/aoc2016/day02/Day02Test.java
package de.havox_design.aoc2016.day02;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Day02Test {

    @SuppressWarnings("java:S2699")
    @Test
    void testMainClass() {
        MainClass.main(new String[0]);
    }

    @ParameterizedTest
    @MethodSource("getDataForPart1")
    void testPart1(String fileName, String expected) {
        Assertions.assertEquals(expected, BathroomSecurity.solvePart1(fileName));
    }

    private static Stream<Arguments> getDataForPart1() {
        return Stream.of(
                Arguments.of("day02Sample.txt", "1985")
        );
    }

    @ParameterizedTest
    @MethodSource("getDataForPart2")
    void testPart2(String fileName, String expected) {
        Assertions.assertEquals(expected, BathroomSecurity.solvePart2(fileName));
    }

    private static Stream<Arguments> getDataForPart2() {
        return Stream.of(
                Arguments.of("day02Sample.txt", "5DB3")
        );
    }
}

File: aoc2022/day02/src/test/kotlin/de/havox_design/aoc2022/day02/Day02Test.kt
package de.havox_design.aoc2022.day02

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

class Day02Test {
    @Test
    fun testMainClass() {
        MainClass.main(arrayOf())
    }

    @ParameterizedTest
    @CsvSource(
        "day02Rock_rock.txt, 4",
        "day02Rock_paper.txt, 8",
        "day02Rock_scissors.txt, 3",
        "day02Paper_rock.txt, 1",
        "day02Paper_paper.txt, 5",
        "day02Paper_scissors.txt, 9",
        "day02Scissors_rock.txt, 7",
        "day02Scissors_paper.txt, 3",
        "day02Scissors_scissors.txt, 6",
        "day02Sample.txt, 15"
    )
    fun testPlaybooks(filename: String, expectedScore: Int) {
        RockPaperScissorsGame(filename)
            .getResultForGuide()
            .shouldBe(expectedScore)
    }

    @ParameterizedTest
    @CsvSource(
        "day02Rock_win.txt, 8",
        "day02Rock_draw.txt, 4",
        "day02Rock_loss.txt, 3",
        "day02Paper_win.txt, 9",
        "day02Paper_draw.txt, 5",
        "day02Paper_loss.txt, 1",
        "day02Scissors_win.txt, 7",
        "day02Scissors_draw.txt, 6",
        "day02Scissors_loss.txt, 2",
        "day02Sample.txt, 12"
    )
    fun testGameByExpectedResults(filename: String, expectedScore: Int) {
        RockPaperScissorsGame(filename)
            .getResultForExpectedResult()
            .shouldBe(expectedScore)
    }
}

private fun Int.shouldBe(expectation: Int) = Assertions.assertEquals(expectation, this)
